src/main/java/net/snowflake/ingest/connection/JWTManager.java

```diff
@@ -0,0 +1,181 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import net.snowflake.ingest.utils.Cryptor;
+
+/**
+ * This class manages creating and automatically renewing the JWT token
+ *
+ * @author obabarinsa
+ * @since 1.8
+ */
+final class JWTManager extends SecurityManager {
+  // the token lifetime is 59 minutes
+  private static final float LIFETIME_IN_MINUTES = 59;
+
+  // the renewal time is 54 minutes
+  private static final int RENEWAL_INTERVAL_IN_MINUTES = 54;
+
+  private static final String TOKEN_TYPE = "KEYPAIR_JWT";
+
+  // The public - private key pair we're using to connect to the service
+  private final transient KeyPair keyPair;
+
+  // the token itself
+  protected final AtomicReference<String> token;
+
+  /**
+   * Creates a JWTManager entity for a given account, user and KeyPair with a specified time to
+   * renew the token
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param keyPair - the public/private key pair we're using to connect
+   * @param timeTillRenewal - the time measure until we renew the token
+   * @param unit the unit by which timeTillRenewal is measured
+   * @param telemetryService reference to the telemetry service
+   */
+  JWTManager(
+      String accountName,
+      String username,
+      KeyPair keyPair,
+      int timeTillRenewal,
+      TimeUnit unit,
+      TelemetryService telemetryService) {
+    super(accountName, username, telemetryService);
+    // if any of our arguments are null, throw an exception
+    if (keyPair == null) {
+      throw new IllegalArgumentException();
+    }
+    token = new AtomicReference<>();
+
+    // we have to keep around the keys
+    this.keyPair = keyPair;
+
+    // generate our first token
+    regenerateToken();
+
+    // schedule all future renewals
+    tokenRefresher.scheduleAtFixedRate(
+        this::regenerateToken, timeTillRenewal, timeTillRenewal, unit);
+  }
+
+  /**
+   * Creates a JWTManager entity for a given account, user and KeyPair with the default time to
+   * renew (RENEWAL_INTERVAL_IN_MINUTES minutes)
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param keyPair - the public/private key pair we're using to connect
+   * @param telemetryService reference to the telemetry service
+   */
+  JWTManager(
+      String accountName, String username, KeyPair keyPair, TelemetryService telemetryService) {
+    this(
+        accountName,
+        username,
+        keyPair,
+        RENEWAL_INTERVAL_IN_MINUTES,
+        TimeUnit.MINUTES,
+        telemetryService);
+  }
+
+  @Override
+  String getToken() {
+    // if we failed to regenerate the token at some point, throw
+    if (refreshFailed.get()) {
+      LOGGER.error("getToken request failed due to token regeneration failure");
+      throw new SecurityException();
```

Comment:
Codecov is complaining about these lines as well.

Reply:
Fixed.